<div style="border: 1px solid thistle; padding: 5px 20px 10px 20px;">
    <h2><font color = "77658f" style=";">Feedback</font></h2><br>
    <font color = "77658f">Lack of problem statement in the project : having a problem statement can help put into context what the project aims to do, the motivation behind building such a model, and how this model can potentially impact stakeholders. Even if this is an exercise, it's important to come up with a context! Good pre-processing and plots, good cleaning of data.<br><br>
        • Problem Statement: <b>Meet Minimum Requirement</b> <br>
        • Data Cleaning and EDA: <b>Meet Requirement</b><br>
        • Preprocessing and Modelling: <b>Meet Requirement</b><br>
        • Evaluation and Conceptual Understanding: <b>Meet Requirement</b><br>
        • Conclusion and Recommendation: <b>Meet Minimum Requirement</b><br>
        • Project Organisation: <b>Meet Requirement</b><br>
        • Visualisations: <b>Meet Requirement</b><br>
        • Python Syntax and Control Flow: <b>Meet Requirement</b><br>
    </font>
</div>

<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Ames Housing - 1 - EDA & Cleaning

_Siti Aishah_

---

### Business Objective
*Create a model that best predicts the price of housing in Ames, Iowa, USA, from past transaction data obtained in 2006-2010*

## 0 Imports

In [1]:
%load_ext autoreload
%autoreload 2

%matplotlib inline

import utility as u #custom functions used across notebooks in this project

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

pd.options.display.max_columns=1000
pd.options.display.max_rows=300

In [2]:
!pwd  ### ???

/Users/sitiaishahahmad/Desktop/DS/Github-GA-DSI/P2_AmesHousing-Modelling/1_notebooks


In [3]:
!ls ../2_data/raw/   ### ???

sample_sub_reg.csv test.csv           train.csv


In [4]:
# Load in training dataset
df = pd.read_csv('../2_data/raw/train.csv')

### Check length of data

In [5]:
len(df)

2051

In [6]:
df

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0      109  533352170           60        RL           NaN     13517   Pave   
1      544  531379050           60        RL          43.0     11492   Pave   
2      153  535304180           20        RL          68.0      7922   Pave   
3      318  916386060           60        RL          73.0      9802   Pave   
4      255  906425045           50        RL          82.0     14235   Pave   
...    ...        ...          ...       ...           ...       ...    ...   
2046  1587  921126030           20        RL          79.0     11449   Pave   
2047   785  905377130           30        RL           NaN     12342   Pave   
2048   916  909253010           50        RL          57.0      7558   Pave   
2049   639  535179160           20        RL          80.0     10400   Pave   
2050    10  527162130           60        RL          60.0      7500   Pave   

     Alley Lot Shape Land Contour Utilities Lot Config Land Slope  \
0      NaN       IR1          Lvl    AllPub    CulDSac        Gtl   
1      NaN       IR1          Lvl    AllPub    CulDSac        Gtl   
2      NaN       Reg          Lvl    AllPub     Inside        Gtl   
3      NaN       Reg          Lvl    AllPub     Inside        Gtl   
4      NaN       IR1          Lvl    AllPub     Inside        Gtl   
...    ...       ...          ...       ...        ...        ...   
2046   NaN       IR1          HLS    AllPub     Inside        Gtl   
2047   NaN       IR1          Lvl    AllPub     Inside        Gtl   
2048   NaN       Reg          Bnk    AllPub     Inside        Gtl   
2049   NaN       Reg          Lvl    AllPub     Corner        Gtl   
2050   NaN       Reg          Lvl    AllPub     Inside        Gtl   

     Neighborhood Condition 1 Condition 2 Bldg Type House Style  Overall Qual  \
0          Sawyer        RRAe        Norm      1Fam      2Story             6   
1         SawyerW        Norm        Norm      1Fam      2Story             7   
2           NAmes        Norm        Norm      1Fam      1Story             5   
3          Timber        Norm        Norm      1Fam      2Story             5   
4         SawyerW        Norm        Norm      1Fam      1.5Fin             6   
...           ...         ...         ...       ...         ...           ...   
2046       Timber        Norm        Norm      1Fam      1Story             8   
2047      Edwards        Norm        Norm      1Fam      1Story             4   
2048      Crawfor        Norm        Norm      1Fam      1.5Fin             6   
2049        NAmes        Norm        Norm      1Fam      1Story             4   
2050      Gilbert        Norm        Norm      1Fam      2Story             7   

      Overall Cond  Year Built  Year Remod/Add Roof Style Roof Matl  \
0                8        1976            2005      Gable   CompShg   
1                5        1996            1997      Gable   CompShg   
2                7        1953            2007      Gable   CompShg   
3                5        2006            2007      Gable   CompShg   
4                8        1900            1993      Gable   CompShg   
...            ...         ...             ...        ...       ...   
2046             5        2007            2007      Gable   CompShg   
2047             5        1940            1950      Gable   CompShg   
2048             6        1928            1950      Gable   CompShg   
2049             5        1956            1956      Gable   CompShg   
2050             5        1999            1999      Gable   CompShg   

     Exterior 1st Exterior 2nd Mas Vnr Type  Mas Vnr Area Exter Qual  \
0         HdBoard      Plywood      BrkFace         289.0         Gd   
1         VinylSd      VinylSd      BrkFace         132.0         Gd   
2         VinylSd      VinylSd         None           0.0         TA   
3         VinylSd      VinylSd         None           0.0         TA   
4         Wd Sdng      Plywood         None 

In [7]:
df.shape

(2051, 81)

## 1. Remove columns with extensive (> 300) missing values 

In [8]:
df.isnull().sum().sort_values(ascending=False).head(25)

Pool QC           2042
Misc Feature      1986
Alley             1911
Fence             1651
Fireplace Qu      1000
Lot Frontage       330
Garage Finish      114
Garage Cond        114
Garage Qual        114
Garage Yr Blt      114
Garage Type        113
Bsmt Exposure       58
BsmtFin Type 2      56
BsmtFin Type 1      55
Bsmt Cond           55
Bsmt Qual           55
Mas Vnr Type        22
Mas Vnr Area        22
Bsmt Half Bath       2
Bsmt Full Bath       2
Garage Cars          1
Garage Area          1
Bsmt Unf SF          1
BsmtFin SF 2         1
Total Bsmt SF        1
dtype: int64

In [9]:
cols_w_extensive_na = df.isnull().sum().sort_values(ascending=False).head(25)

In [10]:
cols_w_extensive_na = cols_w_extensive_na[cols_w_extensive_na > 300]
cols_w_extensive_na = pd.DataFrame(cols_w_extensive_na).T.columns.to_list()
cols_w_extensive_na

['Pool QC', 'Misc Feature', 'Alley', 'Fence', 'Fireplace Qu', 'Lot Frontage']

In [11]:
df.drop(columns = cols_w_extensive_na, inplace=True)

## 2. Remove non-numeric columns whose variation of data is minimal 
(ie. too many of the same category/value/type. threshold = 85%)

In [12]:
# Get numeric columns
numeric_cols = df._get_numeric_data().columns

# Get non-numeric columns
non_numeric_cols = [i for i in df.columns if i not in numeric_cols]

In [13]:
little_info_cols = []

for i in non_numeric_cols:
    
    # Get most no of occurence of a value in the column
    max_cnts = df[i].value_counts().max()
    
    # Get the total number of non-null values in the column
    sum_ = df[i].value_counts().sum()
    
    # If the percentage of max occurence > 85%
    ptg_of_data = max_cnts/sum_
    if ptg_of_data > 0.85:
        little_info_cols.append(i)

little_info_cols

['Street',
 'Land Contour',
 'Utilities',
 'Land Slope',
 'Condition 1',
 'Condition 2',
 'Roof Matl',
 'Exter Cond',
 'Bsmt Cond',
 'BsmtFin Type 2',
 'Heating',
 'Central Air',
 'Electrical',
 'Functional',
 'Garage Qual',
 'Garage Cond',
 'Paved Drive',
 'Sale Type']

In [14]:
# Drop them
df.drop(columns = little_info_cols, inplace=True)

## 3. Remove NUMERIC columns whose variation of data is minimal 
(i.e. too many of the same category/value/type. threshold = 85%)

In [15]:
little_info_cols = []

for i in numeric_cols:
    
    # Get most no of occurence of a value in the column
    max_cnts = df[i].value_counts().max()
    
    # Get the number of non-null values in the column
    sum_ = df[i].value_counts().sum()
    
    # If the percentage of max occurence > 85%
    ptg_of_data = max_cnts/sum_
    if ptg_of_data > 0.85:
        little_info_cols.append(i)

little_info_cols

['BsmtFin SF 2',
 'Low Qual Fin SF',
 'Bsmt Half Bath',
 'Kitchen AbvGr',
 '3Ssn Porch',
 'Screen Porch',
 'Pool Area',
 'Misc Val']

In [16]:
# Drop them
df.drop(little_info_cols, axis=1, inplace=True)

## 4. Check no of NAs in columns

In [17]:
cols_w_na = df.isna().sum()
cols_w_na[cols_w_na > 0].sort_values()

BsmtFin SF 1        1
Bsmt Unf SF         1
Total Bsmt SF       1
Garage Cars         1
Garage Area         1
Bsmt Full Bath      2
Mas Vnr Type       22
Mas Vnr Area       22
Bsmt Qual          55
BsmtFin Type 1     55
Bsmt Exposure      58
Garage Type       113
Garage Yr Blt     114
Garage Finish     114
dtype: int64

##### Since theres not many NAs in these columns, we remove the rows that has NA values in these columns instead.

In [18]:
no_of_rows_with_na = len(df[df.isnull().any(axis=1)])
no_of_rows_with_na

184

##### There are only 184 rows with NA; the number is small so it is safe to remove them.

In [19]:
df.dropna(inplace=True)

In [20]:
df.shape

(1867, 49)

## 5. Eyeball Observation/Intuition/Common Sense/Much Reading of data doc tells us that some columns are related with one other

In [21]:
df.head(3)

Id        PID  MS SubClass MS Zoning  Lot Area Lot Shape Lot Config  \
0  109  533352170           60        RL     13517       IR1    CulDSac   
1  544  531379050           60        RL     11492       IR1    CulDSac   
2  153  535304180           20        RL      7922       Reg     Inside   

  Neighborhood Bldg Type House Style  Overall Qual  Overall Cond  Year Built  \
0       Sawyer      1Fam      2Story             6             8        1976   
1      SawyerW      1Fam      2Story             7             5        1996   
2        NAmes      1Fam      1Story             5             7        1953   

   Year Remod/Add Roof Style Exterior 1st Exterior 2nd Mas Vnr Type  \
0            2005      Gable      HdBoard      Plywood      BrkFace   
1            1997      Gable      VinylSd      VinylSd      BrkFace   
2            2007      Gable      VinylSd      VinylSd         None   

   Mas Vnr Area Exter Qual Foundation Bsmt Qual Bsmt Exposure BsmtFin Type 1  \
0         289.0         Gd     CBlock        TA            No            GLQ   
1         132.0         Gd      PConc        Gd            No            GLQ   
2           0.0         TA     CBlock        TA            No            GLQ   

   BsmtFin SF 1  Bsmt Unf SF  Total Bsmt SF Heating QC  1st Flr SF  \
0         533.0        192.0          725.0         Ex         725   
1         637.0        276.0          913.0         Ex         913   
2         731.0        326.0         1057.0         TA        1057   

   2nd Flr SF  Gr Liv Area  Bsmt Full Bath  Full Bath  Half Bath  \
0         754         1479             0.0          2          1   
1        1209         2122             1.0          2          1   
2           0         1057             1.0          1          0   

   Bedroom AbvGr Kitchen Qual  TotRms AbvGrd  Fireplaces Garage Type  \
0              3           Gd              6           0      Attchd   
1              4           Gd              8           1      Attchd   
2              3           Gd              5           0      Detchd   

   Garage Yr Blt Garage Finish  Garage Cars  Garage Area  Wood Deck SF  \
0         1976.0           RFn          2.0        475.0             0   
1         1997.0           RFn          2.0        559.0             0   
2         1953.0           Unf          1.0        246.0             0   

   Open Porch SF  Enclosed Porch  Mo Sold  Yr Sold  SalePrice  
0             44               0        3     2010     130500  
1             74               0        4     2009     220000  
2             52               0        1     2010     109000

### Remove `ID` since there is `PID` as another unique identifier

In [22]:
df.PID.drop_duplicates()

0       533352170
1       531379050
2       535304180
3       916386060
4       906425045
          ...    
2046    921126030
2047    905377130
2048    909253010
2049    535179160
2050    527162130
Name: PID, Length: 1867, dtype: int64

In [23]:
df.drop(columns="Id", inplace=True)

### `BsmtFin SF 1` + `Bsmt Unf SF` = `Total Bsmt SF	`

In [24]:
df['checks'] = df.apply(lambda x: 1 if (x['BsmtFin SF 1'] + x['Bsmt Unf SF']) == x['Total Bsmt SF'] else 0, axis=1)

In [25]:
round(len(df[df.checks == 1]) / len(df) * 100, 1)

87.0

- *Even though some rows have sum not equals to 'Total Bsmt SF', 87%. of them are. We decide keep 'Total Bsmt SF' being more holistic, and only 'BsmtFin SF 1'*

In [26]:
df.drop(columns='Bsmt Unf SF', inplace=True)

### `1st Flr SF` + `2nd Flr SF` = `Gr Liv Area`

In [27]:
df['checks'] = df.apply(lambda x: 1 if (x['1st Flr SF'] + x['2nd Flr SF']) == x['Gr Liv Area'] else 0, axis=1)

In [28]:
round(len(df[df.checks == 1]) / len(df) * 100, 1)

98.8

- *Even though there are rows with sum not equals to 'Gr Liv Area', 98.8%. of them are. We decide to drop '2nd Flr SF' to remove its explicit relationship with 'Gr Liv Area'.*

In [29]:
df.drop(columns='2nd Flr SF', inplace=True)

### `Year Built` vs `Year Remod/Add`

In the description:

>Year Remod/Add (Discrete): Remodel date (same as construction date if no remodeling or additions)

- *Since as buyers, we would care more about how recent a remodelling or addition has taken place, we drop `Year Built`*

In [30]:
df.drop(columns='Year Built', inplace=True)

### Columns such as `Exterior 1st`, `Exterior 2nd`, `Mas Vnr Type` are hedonic factors but can objectively be evaluated using `Exter Qual`. 

### Similarly, `Bsmt Exposure`, `BsmtFin Type 1` can be evaluated by `Bsmt Qual`

In [31]:
df.drop(columns=[ 'Exterior 1st', 'Exterior 2nd', 'Mas Vnr Type', 'Bsmt Exposure', 'BsmtFin Type 1'], inplace=True)

### `MS Subclass` similar to `House Style`

In the description:

>House Style (Nominal): Style of dwelling
	
       1Story	One story
       1.5Fin	One and one-half story: 2nd level finished
       1.5Unf	One and one-half story: 2nd level unfinished
       2Story	Two story
       2.5Fin	Two and one-half story: 2nd level finished
       2.5Unf	Two and one-half story: 2nd level unfinished
       SFoyer	Split Foyer
       SLvl	Split Level
       
>MS SubClass (Nominal): Identifies the type of dwelling involved in the sale.	

       020	1-STORY 1946 & NEWER ALL STYLES
       030	1-STORY 1945 & OLDER
       040	1-STORY W/FINISHED ATTIC ALL AGES
       045	1-1/2 STORY - UNFINISHED ALL AGES
       050	1-1/2 STORY FINISHED ALL AGES
       060	2-STORY 1946 & NEWER
       070	2-STORY 1945 & OLDER
       075	2-1/2 STORY ALL AGES
       080	SPLIT OR MULTI-LEVEL
       085	SPLIT FOYER
       090	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES

- *For simplicity, we will keep House Style and drop MS SubClass's with a little too many descriptions*

In [32]:
df.drop(columns='MS SubClass', inplace=True)

## EXPORT for use in Feat Eng (Notebook 2)

In [33]:
df.drop(columns='checks', inplace=True)

In [34]:
df.shape #we're left with 1867 entries and 39 columns (1st is PID, last is SalePrice)

(1867, 39)

In [35]:
# parquet allows one to keep the data as is, compared to csv which might cause some losses in formatting when importing later

df.to_parquet('../2_data/interim/cleaned.parquet')